multinet/__init__.py

```diff
@@ -21,10 +21,21 @@
 sentry_sdk.init(dsn=sentry_dsn, integrations=[FlaskIntegration()])
 
 
+def get_cors_hosts() -> List[str]:
+    """Read in comma-separated CORS hosts list from environment."""
```

Comment:
It's actually the list of domains that the server will accept cookies from. I am not actually sure if we need to explicitly list the individual apps' domains or if adding `https://multinet.app` will cover those.

Reply:
See 5702fb0c279c5ab116f16d947fce9ec9473f3215.

Reply:
This might just need a regex. From the [documentation](https://flask-cors.corydolphin.com/en/latest/api.html#flask_cors.CORS):

>The origin, or list of origins to allow requests from. The origin(s) may be regular expressions, case-sensitive strings, or else an asterisk

Reply:
> It's actually the list of domains that the server will accept cookies from.

It's both, isn't it? I mean, doesn't this list specify which domains we allow requests to be made from? We just happen to allow cookies to be passed from these domains as well.

Reply:
> This might just need a regex. From the [documentation](https://flask-cors.corydolphin.com/en/latest/api.html#flask_cors.CORS):
> 
> > The origin, or list of origins to allow requests from. The origin(s) may be regular expressions, case-sensitive strings, or else an asterisk

Good call. I will do some local experimentation to figure out exactly how this works. Something like `https://.*\.multinet\.app` (note the backslashes) might be the right answer here.

Reply:
> > It's actually the list of domains that the server will accept cookies from.
> 
> It's both, isn't it? I mean, doesn't this list specify which domains we allow requests to be made from? We just happen to allow cookies to be passed from these domains as well.

Yeah, you are right, sorry for the confusion. The problem is, I am just not sure how to specify both of the following:
1. a whitelist of domains allowed to send cookies (`supports_credentials = True`)
2. that all hosts may ask for data (CORS headers set to `*`)

If we can do this using that `after_request` hook, I'm down to trying that out (in a followup PR).

Let's talk about this on Monday.

Reply:
> Yeah, you are right, sorry for the confusion. The problem is, I am just not sure how to specify both of the following:
> 
> 1. a whitelist of domains allowed to send cookies (`supports_credentials = True`)
> 2. that all hosts may ask for data (CORS headers set to `*`)

I'm not sure about this (haven't tested locally), but this configuration with Flask CORS may work:

```
allowed_origins = get_allowed_origins()
CORS(
    app,
    resources={
        "/api/user/.*": {"origins": allowed_origins, "supports-credentials": True},
        "/api/.*": {"origins": "*"},
    },
)
```

My hope is that the first one takes precedence over the second one, kind of how `vue-router` routes work on the client. The first resource would cover all of our current user endpoints, which is logging in, getting info, and logging out. In the future we could also probably generalize this with a simple list of endpoints that we just convert into a regex before calling this function. Something like:

```
access_controlled_endpoints = "|".join([
    r"^/api/user/.*$",
    r"^/api/some_other_endpoint$",
])
```

But this is mainly speculation. I'm not even sure Flask CORS handles precedence like I think it does. There's also options to use the `CORS` function on individual blueprints, as well as the [`cross_origin`](https://flask-cors.corydolphin.com/en/latest/api.html#flask_cors.cross_origin) decorator.

Reply:
This may well be a valid configuration, but it won't work for our purposes. The problem is, we really do need for clients to send cookies for more than just the `/user` endpoints. To invoke the example I've been using, if a client is logged in as `waxlamp`, and sends a `GET /workspaces` request under your proposed scheme, the response would only contain the public workspaces and not non-public ones readable by `waxlamp`.

So the more I think about this, the more I believe our possibilities are:
1. Maintain a whitelist of allowed origins (this is what I would informally call the "Google approach").
2. Make the login endpoints publicly accessible, and design them to return a token, which clients manually stuff into the header when they make requests (the "Girder approach").

(2) is pretty involved, IMO. We can follow Girder's lead, but we'd be partially rolling our own login infrastructure, and we'd be complicating a general client's ability to get to our server (since they have to either manually prepare a custom header in their requests, or else only use our libraries to do so). All this seems to contraindicate our goal of deferring all this stuff to the relative ease of using OAuth in the first place.

We might also consider talking to the Girder team about what they've done and what they've learned on this topic.

(1) seems to be a fine approach to me, seeing as we're a research project, it's not difficult to add custom origins, and it's a simpler overall solution.

The one big issue with (1) is that it ends up forbidding using the production server for local client testing. In the "real world", testing with production would of course be non-kosher, but again: we are a research project. So maybe there's a compromise: perhaps we include `http://localhost:<some high port number>` in the whitelist, then shift our dev environments to build and serve the client on that high port.

Let's discuss this stuff in person and we'll come to a decision.

Reply:
In our meeting we decided to stick with the whitelist-based approach for now, and investigate something more flexible if time allows and there is sufficient need/demand for it. Issue #384 tracks that future task.